src/voting_nodes.rs

```diff
@@ -0,0 +1,352 @@
+//! The `voting_nodes` maintains a map of peer nodes that are actively
+//! voting on the transactions
+
+use byteorder::{LittleEndian, ReadBytesExt};
+use counter::Counter;
+use crdt::{Crdt, NodeInfo};
+use hash::Hash;
+use log::Level;
+use signature::Pubkey;
+use std::collections::HashMap;
+use std::io::Cursor;
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::{Arc, RwLock};
+use std::thread::{sleep, Builder};
+use std::time::Duration;
+use timing::timestamp;
+use transaction::Vote;
+
+const VOTING_NODES_PURGE_SLEEP_MILLIS: u64 = 100;
+const VOTING_NODES_PURGE_MILLIS: u64 = 15000;
+
+pub struct VotingNodes {
+    pub nodes: Arc<RwLock<HashMap<Pubkey, u64>>>,
+}
+
+fn make_debug_id(key: &Pubkey) -> u64 {
+    let buf: &[u8] = &key.as_ref();
+    let mut rdr = Cursor::new(&buf[..8]);
+    rdr.read_u64::<LittleEndian>()
+        .expect("rdr.read_u64 in fn debug_id")
+}
+
+impl VotingNodes {
+    pub fn new(crdt: Arc<RwLock<Crdt>>, exit: Arc<AtomicBool>) -> Self {
+        let nodes_arc = Arc::new(RwLock::new(HashMap::new()));
+        let me = VotingNodes {
+            nodes: nodes_arc.clone(),
+        };
+
+        Builder::new()
+            .name("solana-voting-nodes".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    return;
+                }
+                VotingNodes::purge(&crdt, &nodes_arc, timestamp());
```

Comment:
Instead of having a thread actively maintaining this hash map, how about calling it from the methods that expect it to have been called? That is, call it from `compute_broadcast_table`? Seems like that'd simplify code in other places and may allow you to toss the `Arc<RwLock>` wrappers.

Reply:
Makes sense. It'll simplify the code.